{"cells":[{"cell_type":"markdown","id":"f91a22cc-4d57-4355-a616-c5eca0f9b70b","metadata":{"id":"f91a22cc-4d57-4355-a616-c5eca0f9b70b"},"source":["# Instructions\n","\n","This problem set is provided in the form of a Jupyter notebook. Problems are posed within this notebook file and you are expected to provide codes and/or written answers when prompted. Remember that you can use Markdown cells to format written responses where necessary.\n","\n","Before submitting your assignment, be sure to do a clean run of your notebook and **verify that your cell outputs (e.g., prints, figures, tables) are correctly shown**. To do a clean run, click *Kernel&#8594;Restart Kernal and Run All Cells...*.\n","\n","You are required to submit this notebook to Gradescope in two forms:\n","\n","1. Submit a PDF of the completed notebook. To produce a PDF, you can use *File&#8594;Save and Export Notebook As...&#8594;HTML* and then convert the HTML file to a PDF using your preferred web browser. **Verify that your code, written answers, and cell outputs are visible in the submitted PDF.**\n","2. Submit a zip file (including the `.ipynb` file) of this assignment to Gradescope."]},{"cell_type":"markdown","id":"4e20e104-5c5c-4ee6-b3c0-d74bd02defbc","metadata":{"id":"4e20e104-5c5c-4ee6-b3c0-d74bd02defbc"},"source":["# Online Resources and Collaborators\n","\n","Please list the online resources you used and the names of other students you collaborated with while working on this problem set.\n","* Online resoures: https://www.biostars.org/p/19122/#:~:text=high%20gap%20opening%20penalty:%20point,aligned%20across%20their%20full%20length.)\n","* Student Names:"]},{"cell_type":"markdown","id":"7755b53f-6fc2-4ca1-b96a-24a980a941fa","metadata":{"id":"7755b53f-6fc2-4ca1-b96a-24a980a941fa"},"source":["# Problem set 0: introduction to genomic data.\n","\n","In this problem set, you will access the genome of the original Sars-CoV-2 strain (causal agent of the covid-19 pandemic) and translate one of its genes into a protein product. The learning objectives are:\n","1. Become familiar with BioPython\n","2. Become familiar with fasta formatted sequence files\n","3. Build understanding of how information from DNA is translated into protein by implementing your own translation function"]},{"cell_type":"markdown","id":"bdfb5308-76ed-4a1f-b591-e39d1a564db0","metadata":{"id":"bdfb5308-76ed-4a1f-b591-e39d1a564db0"},"source":["## 00. Install BioPython (5 points)\n","\n","You can find the installation instructions here: https://biopython.org/wiki/Download"]},{"cell_type":"code","execution_count":144,"id":"a8a1d439-a555-411a-af3c-a29f7896f21a","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":12793,"status":"ok","timestamp":1770762090979,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"a8a1d439-a555-411a-af3c-a29f7896f21a","outputId":"85daeb30-e5ca-4e41-9639-7be05e787cde"},"outputs":[{"output_type":"stream","name":"stdout","text":["Requirement already satisfied: biopython in /usr/local/lib/python3.12/dist-packages (1.86)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n"]}],"source":["# install BioPython and import it\n","!pip install biopython\n","\n","from Bio import SeqIO\n","import pandas as pd"]},{"cell_type":"code","execution_count":145,"id":"91esnE4KTgDv","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":833,"status":"ok","timestamp":1770762091814,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"91esnE4KTgDv","outputId":"53437a67-bd47-4a7f-9c32-448bd0a45fae"},"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n","Archive:  /content/drive/MyDrive/690U_assignments/ncbi_dataset.zip\n","  inflating: /content/README.md      \n","  inflating: /content/ncbi_dataset/data/data_report.jsonl  \n","  inflating: /content/ncbi_dataset/data/genomic.fna  \n","  inflating: /content/ncbi_dataset/data/protein.faa  \n","  inflating: /content/ncbi_dataset/data/virus_dataset.md  \n","  inflating: /content/ncbi_dataset/data/dataset_catalog.json  \n","  inflating: /content/md5sum.txt     \n","data_report.jsonl     genomic.fna  virus_dataset.md\n","dataset_catalog.json  protein.faa\n"]}],"source":["from google.colab import drive\n","drive.mount('/content/drive')\n","\n","# Unzip the ncbi_dataset.zip file into the Colab environment's filesystem (in-memory)\n","!unzip -o '/content/drive/MyDrive/690U_assignments/ncbi_dataset.zip' -d '/content/'\n","\n","# Optional: List the contents of the unzipped directory to verify\n","!ls '/content/ncbi_dataset/data/'"]},{"cell_type":"code","execution_count":146,"id":"DGPYdljbUQkF","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":141,"status":"ok","timestamp":1770762091956,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"DGPYdljbUQkF","outputId":"a106beba-eef4-4651-bd66-51db5f65ddb9"},"outputs":[{"output_type":"stream","name":"stdout","text":["drive/\tmd5sum.txt  ncbi_dataset/  README.md  sample_data/\n"]}],"source":["!ls -F /content"]},{"cell_type":"markdown","id":"2ade6b73-d252-4555-b01a-44b44495b5b6","metadata":{"id":"2ade6b73-d252-4555-b01a-44b44495b5b6"},"source":["## 01. Use Bio.SeqIO to read in the genome sequence of Sars-CoV-2 (15 points)"]},{"cell_type":"markdown","id":"2cd14fa7-bfbd-4465-a255-c1346d544c5e","metadata":{"id":"2cd14fa7-bfbd-4465-a255-c1346d544c5e"},"source":["The covid-19 pandemic was caused by a virus called Sars-Cov-2. One of the original Sars-CoV-2 genomes sequenced is this one: https://www.ncbi.nlm.nih.gov/datasets/taxonomy/2697049/\n","\n","Use the \"download\" button to access the genome sequence and the protein sequence data from this genome. You can view a graphical layout of the genes in the genome on the website as well. DO NOT click the \"download all genomes\" button! This downloads all the viral sequences on NCBI\n","\n","Use Bio.SeqIO to read in the fasta-formatted **genomic** data file that you downloaded. This should contain a single nucleotide sequence, representing the full genome\n","\n","SeqIO documentation: https://biopython.org/docs/1.75/api/Bio.SeqIO.html (hint: use the 'parse' function)\n"]},{"cell_type":"code","execution_count":147,"id":"ldTO7wJVYboa","metadata":{"collapsed":true,"executionInfo":{"elapsed":1,"status":"ok","timestamp":1770762091969,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"ldTO7wJVYboa"},"outputs":[],"source":["# for protein_record in SeqIO.parse(\"/content/ncbi_dataset/data/protein.faa\", \"fasta\"):\n","#     print(\"%s %i\" % (protein_record.id, len(protein_record)))"]},{"cell_type":"code","execution_count":148,"id":"95fab142-9ef6-4b45-823e-30c03d67a6f0","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":53,"status":"ok","timestamp":1770762092033,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"95fab142-9ef6-4b45-823e-30c03d67a6f0","outputId":"7d1fbd91-6a08-4d24-f9ae-f096ace4f3d9"},"outputs":[{"output_type":"stream","name":"stdout","text":["NC_045512.2 29903\n","Gene:  ID: NC_045512.2\n","Name: NC_045512.2\n","Description: NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n","Number of features: 0\n","Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA')\n"]}],"source":["for gene_record in SeqIO.parse(\"/content/ncbi_dataset/data/genomic.fna\", \"fasta\"):\n","    print(\"%s %i\" % (gene_record.id, len(gene_record)))\n","    print(\"Gene: \", gene_record)"]},{"cell_type":"markdown","id":"360d51d8-ad85-4731-91ea-446f780973a0","metadata":{"id":"360d51d8-ad85-4731-91ea-446f780973a0"},"source":["## 02. Translate the sequence of the Spike protein (25 points total)"]},{"cell_type":"markdown","id":"5365ea0d-0663-4d97-9baf-16873f9394c4","metadata":{"id":"5365ea0d-0663-4d97-9baf-16873f9394c4"},"source":["The spike protein is the protein responsible for facilitating entry of covid-19 into human cells.  It is also the protein targeted by antibodies produced by our immune system. Thus, mutations in the spike protein are extremely interesting - they may facilitate increased ability of the virus to evade immunity or enter cells. We are going to translate the gene sequence of the spike protein (the gene is called \"S\") into a protein sequence.  This gene is encoded from position *21563* to *25384* (1-indexed) on the chromosome in the + direction (you can see info on this gene here: https://www.ncbi.nlm.nih.gov/gene/43740568/).\n"]},{"cell_type":"markdown","id":"cae80fba-5d71-4a8e-bbba-5c0a28fc65a8","metadata":{"id":"cae80fba-5d71-4a8e-bbba-5c0a28fc65a8"},"source":["### 02.A extract the gene sequence (subsequence of the genome sequence) corresponding to the spike protein (10 points)"]},{"cell_type":"code","execution_count":149,"id":"5273cc32-a4f6-472f-8b73-9fe034ac740c","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1770762092037,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"5273cc32-a4f6-472f-8b73-9fe034ac740c","outputId":"ef32446d-6488-4011-dea4-38eeb72c204d"},"outputs":[{"output_type":"stream","name":"stdout","text":["Spike gene length (nt): 3822\n","ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTGTTAATCTTACAACCAGAACTCAATTACCCCCTGCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTTTTCAGATCCTCAGTTTTACATTCAACTCAGGACTTGTTCTTACCTTTCTTTTCCAATGTTACTTGGTTCCATGCTATACATGTCTCTGGGACCAATGGTACTAAGAGGTTTGATAACCCTGTCCTACCATTTAATGATGGTGTTTATTTTGCTTCCACTGAGAAGTCTAACATAATAAGAGGCTGGATTTTTGGTACTACTTTAGATTCGAAGACCCAGTCCCTACTTATTGTTAATAACGCTACTAATGTTGTTATTAAAGTCTGTGAATTTCAATTTTGTAATGATCCATTTTTGGGTGTTTATTACCACAAAAACAACAAAAGTTGGATGGAAAGTGAGTTCAGAGTTTATTCTAGTGCGAATAATTGCACTTTTGAATATGTCTCTCAGCCTTTTCTTATGGACCTTGAAGGAAAACAGGGTAATTTCAAAAATCTTAGGGAATTTGTGTTTAAGAATATTGATGGTTATTTTAAAATATATTCTAAGCACACGCCTATTAATTTAGTGCGTGATCTCCCTCAGGGTTTTTCGGCTTTAGAACCATTGGTAGATTTGCCAATAGGTATTAACATCACTAGGTTTCAAACTTTACTTGCTTTACATAGAAGTTATTTGACTCCTGGTGATTCTTCTTCAGGTTGGACAGCTGGTGCTGCAGCTTATTATGTGGGTTATCTTCAACCTAGGACTTTTCTATTAAAATATAATGAAAATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCTCAGAAACAAAGTGTACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACTTGTGCCCTTTTGGTGAAGTTTTTAACGCCACCAGATTTGCATCTGTTTATGCTTGGAACAGGAAGAGAATCAGCAACTGTGTTGCTGATTATTCTGTCCTATATAATTCCGCATCATTTTCCACTTTTAAGTGTTATGGAGTGTCTCCTACTAAATTAAATGATCTCTGCTTTACTAATGTCTATGCAGATTCATTTGTAATTAGAGGTGATGAAGTCAGACAAATCGCTCCAGGGCAAACTGGAAAGATTGCTGATTATAATTATAAATTACCAGATGATTTTACAGGCTGCGTTATAGCTTGGAATTCTAACAATCTTGATTCTAAGGTTGGTGGTAATTATAATTACCTGTATAGATTGTTTAGGAAGTCTAATCTCAAACCTTTTGAGAGAGATATTTCAACTGAAATCTATCAGGCCGGTAGCACACCTTGTAATGGTGTTGAAGGTTTTAATTGTTACTTTCCTTTACAATCATATGGTTTCCAACCCACTAATGGTGTTGGTTACCAACCATACAGAGTAGTAGTACTTTCTTTTGAACTTCTACATGCACCAGCAACTGTTTGTGGACCTAAAAAGTCTACTAATTTGGTTAAAAACAAATGTGTCAATTTCAACTTCAATGGTTTAACAGGCACAGGTGTTCTTACTGAGTCTAACAAAAAGTTTCTGCCTTTCCAACAATTTGGCAGAGACATTGCTGACACTACTGATGCTGTCCGTGATCCACAGACACTTGAGATTCTTGACATTACACCATGTTCTTTTGGTGGTGTCAGTGTTATAACACCAGGAACAAATACTTCTAACCAGGTTGCTGTTCTTTATCAGGATGTTAACTGCACAGAAGTCCCTGTTGCTATTCATGCAGATCAACTTACTCCTACTTGGCGTGTTTATTCTACAGGTTCTAATGTTTTTCAAACACGTGCAGGCTGTTTAATAGGGGCTGAACATGTCAACAACTCATATGAGTGTGACATACCCATTGGTGCAGGTATATGCGCTAGTTATCAGACTCAGACTAATTCTCCTCGGCGGGCACGTAGTGTAGCTAGTCAATCCATCATTGCCTACACTATGTCACTTGGTGCAGAAAATTCAGTTGCTTACTCTAATAACTCTATTGCCATACCCACAAATTTTACTATTAGTGTTACCACAGAAATTCTACCAGTGTCTATGACCAAGACATCAGTAGATTGTACAATGTACATTTGTGGTGATTCAACTGAATGCAGCAATCTTTTGTTGCAATATGGCAGTTTTTGTACACAATTAAACCGTGCTTTAACTGGAATAGCTGTTGAACAAGACAAAAACACCCAAGAAGTTTTTGCACAAGTCAAACAAATTTACAAAACACCACCAATTAAAGATTTTGGTGGTTTTAATTTTTCACAAATATTACCAGATCCATCAAAACCAAGCAAGAGGTCATTTATTGAAGATCTACTTTTCAACAAAGTGACACTTGCAGATGCTGGCTTCATCAAACAATATGGTGATTGCCTTGGTGATATTGCTGCTAGAGACCTCATTTGTGCACAAAAGTTTAACGGCCTTACTGTTTTGCCACCTTTGCTCACAGATGAAATGATTGCTCAATACACTTCTGCACTGTTAGCGGGTACAATCACTTCTGGTTGGACCTTTGGTGCAGGTGCTGCATTACAAATACCATTTGCTATGCAAATGGCTTATAGGTTTAATGGTATTGGAGTTACACAGAATGTTCTCTATGAGAACCAAAAATTGATTGCCAACCAATTTAATAGTGCTATTGGCAAAATTCAAGACTCACTTTCTTCCACAGCAAGTGCACTTGGAAAACTTCAAGATGTGGTCAACCAAAATGCACAAGCTTTAAACACGCTTGTTAAACAACTTAGCTCCAATTTTGGTGCAATTTCAAGTGTTTTAAATGATATCCTTTCACGTCTTGACAAAGTTGAGGCTGAAGTGCAAATTGATAGGTTGATCACAGGCAGACTTCAAAGTTTGCAGACATATGTGACTCAACAATTAATTAGAGCTGCAGAAATCAGAGCTTCTGCTAATCTTGCTGCTACTAAAATGTCAGAGTGTGTACTTGGACAATCAAAAAGAGTTGATTTTTGTGGAAAGGGCTATCATCTTATGTCCTTCCCTCAGTCAGCACCTCATGGTGTAGTCTTCTTGCATGTGACTTATGTCCCTGCACAAGAAAAGAACTTCACAACTGCTCCTGCCATTTGTCATGATGGAAAAGCACACTTTCCTCGTGAAGGTGTCTTTGTTTCAAATGGCACACACTGGTTTGTAACACAAAGGAATTTTTATGAACCACAAATCATTACTACAGACAACACATTTGTGTCTGGTAACTGTGATGTTGTAATAGGAATTGTCAACAACACAGTTTATGATCCTTTGCAACCTGAATTAGACTCATTCAAGGAGGAGTTAGATAAATATTTTAAGAATCATACATCACCAGATGTTGATTTAGGTGACATCTCTGGCATTAATGCTTCAGTTGTAAACATTCAAAAAGAAATTGACCGCCTCAATGAGGTTGCCAAGAATTTAAATGAATCTCTCATCGATCTCCAAGAACTTGGAAAGTATGAGCAGTATATAAAATGGCCATGGTACATTTGGCTAGGTTTTATAGCTGGCTTGATTGCCATAGTAATGGTGACAATTATGCTTTGCTGTATGACCAGTTGCTGTAGTTGTCTCAAGGGCTGTTGTTCTTGTGGATCCTGCTGCAAATTTGATGAAGACGACTCTGAGCCAGTGCTCAAAGGAGTCAAATTACATTACACATAA\n"]}],"source":["spike_start = 21562\n","spike_end = 25384\n","spike_seq = gene_record.seq[spike_start:spike_end]\n","\n","print(\"Spike gene length (nt):\", len(spike_seq))\n","print(spike_seq)"]},{"cell_type":"markdown","id":"4d542cc7-2395-4963-abb9-fd3bd4e4e038","metadata":{"id":"4d542cc7-2395-4963-abb9-fd3bd4e4e038"},"source":["### 02.B Use the translation table below to translate the sequence of the spike gene (\"S\") into a protein sequence (15 points)"]},{"cell_type":"code","execution_count":150,"id":"1e7f8b88-6030-40fe-831a-2e295499362a","metadata":{"id":"1e7f8b88-6030-40fe-831a-2e295499362a","executionInfo":{"status":"ok","timestamp":1770762092045,"user_tz":300,"elapsed":7,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"}}},"outputs":[],"source":["# Use the translation table\n","# _ indicates a stop codon\n","translation_table ={\n","   \"TTT\":\"F\", \"TTC\":\"F\", \"TTA\":\"L\", \"TTG\":\"L\",\n","   \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\",\n","   \"TAT\":\"Y\", \"TAC\":\"Y\", \"TAA\":\"_\", \"TAG\":\"_\",\n","   \"TGT\":\"C\", \"TGC\":\"C\", \"TGA\":\"_\", \"TGG\":\"W\",\n","   \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n","   \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n","   \"CAT\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n","   \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n","   \"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"I\", \"ATG\":\"M\",\n","   \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n","   \"AAT\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n","   \"AGT\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n","   \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n","   \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n","   \"GAT\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n","   \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\"\n","}"]},{"cell_type":"code","execution_count":151,"id":"v3Vq24yWw4ge","metadata":{"id":"v3Vq24yWw4ge","executionInfo":{"status":"ok","timestamp":1770762092072,"user_tz":300,"elapsed":12,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"}}},"outputs":[],"source":["def translate_seq(nuc_seq, translation_table, unknown='X'):\n","  protein_seq = \"\"\n","  for i in range(0, len(spike_seq)-2, 3):\n","    codon = str(spike_seq[i:i+3])\n","    translation = translation_table.get(codon, \"X\")\n","    protein_seq += translation\n","  return protein_seq, len(protein_seq)"]},{"cell_type":"code","execution_count":152,"id":"99797983-aefa-4e24-ad68-6135b534252c","metadata":{"id":"99797983-aefa-4e24-ad68-6135b534252c","executionInfo":{"status":"ok","timestamp":1770762092077,"user_tz":300,"elapsed":3,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"}}},"outputs":[],"source":["# translate the sequence of the spike gene into a protein sequence and print it\n","\n","protein_seq = translate_seq(spike_seq, translation_table)[0]"]},{"cell_type":"markdown","id":"d8c1f298-9a3d-4b89-8fec-fca235123ca7","metadata":{"id":"d8c1f298-9a3d-4b89-8fec-fca235123ca7"},"source":["## 03. Check your work (5 points)"]},{"cell_type":"markdown","id":"92612055-f8a5-4e5f-a4c8-227090ca50ce","metadata":{"id":"92612055-f8a5-4e5f-a4c8-227090ca50ce"},"source":["The protein.faa file you downloaded contains the translated sequences of all genes in the genome. The identifier (\"id\") for the S protein is called **YP_009724390.1:1-1273**. Again using Bio.SeqIO, read in the protein sequences from the protein.faa file, select the sequence corresponding to the spike protein based on its identifier, and compare it to your translated sequence. They should match, except that our translation table above puts a \"\\_\" at the end of sequences to represent the stop codon!\n","\n","I realize that comparisons like this can be somewhat boring, but it's very important to build checks into your pipelines! Biological data can be very heterogenous."]},{"cell_type":"code","execution_count":153,"id":"ba07bb98-dbe3-41fe-9b6d-be4eac7c7f6c","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":6,"status":"ok","timestamp":1770762092084,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"ba07bb98-dbe3-41fe-9b6d-be4eac7c7f6c","outputId":"d6b576cb-a3b8-4f27-c36e-c180ede6eb4b"},"outputs":[{"output_type":"stream","name":"stdout","text":["Length of reference spike: 1273\n","Reference spike: MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAIHVSGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLGVYYHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINLVRDLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPCNGVEGFNCYFPLQSYGFQPTNGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIADTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQDVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSYECDIPIGAGICASYQTQTNSPRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT\n"]}],"source":["for record in SeqIO.parse(\"/content/ncbi_dataset/data/protein.faa\", \"fasta\"):\n","  if record.id.startswith(\"YP_009724390.1:1-1273\"):\n","          reference_spike = str(record.seq)\n","          break\n","\n","print(\"Length of reference spike:\", len(reference_spike))\n","print(\"Reference spike:\", reference_spike)"]},{"cell_type":"code","execution_count":154,"id":"LPqtTL403IUf","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":52,"status":"ok","timestamp":1770762092137,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"LPqtTL403IUf","outputId":"cc65f0e6-140c-40ec-97a0-8f7a97452c97"},"outputs":[{"output_type":"stream","name":"stdout","text":["Matches: True\n"]}],"source":["matches = protein_seq[:-1] == reference_spike\n","print(\"Matches:\", matches)"]},{"cell_type":"markdown","id":"bb9972ef-9177-4f64-999f-000523d3d1b5","metadata":{"id":"bb9972ef-9177-4f64-999f-000523d3d1b5"},"source":["## 04. Implement the Needleman-Wunsch algorithm and interpret (45 points total)"]},{"cell_type":"markdown","id":"b4274260-b3ed-49de-8607-2c6e7956d018","metadata":{"id":"b4274260-b3ed-49de-8607-2c6e7956d018"},"source":["### 04A. Implement the Needleman-Wunsch algorithm using the BLOSUM80 substitution matrix. (25 points)"]},{"cell_type":"markdown","id":"abceb924-d706-4fc0-90bf-c919fae40aba","metadata":{"id":"abceb924-d706-4fc0-90bf-c919fae40aba"},"source":["\n","I suggest using the blosum python package which provides the substitution matrices https://github.com/not-a-feature/blosum\n","\n","Use a gap penalty of -8. Provide an alignment for the two sequences below, as well as their score:"]},{"cell_type":"code","execution_count":157,"id":"9f35b3f9-469b-489c-9cc6-7dcea0f4e304","metadata":{"id":"9f35b3f9-469b-489c-9cc6-7dcea0f4e304","colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"status":"ok","timestamp":1770762311319,"user_tz":300,"elapsed":6742,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"}},"outputId":"b2709e39-2593-4048-cdce-8151fd1e5717"},"outputs":[{"output_type":"stream","name":"stdout","text":["Requirement already satisfied: blosum in /usr/local/lib/python3.12/dist-packages (2.2.0)\n"]}],"source":["!pip install blosum\n","\n","import blosum as bl\n","import numpy as np\n","\n","def NW(sequence1, sequence2, S, gap_penalty=-8):\n","    \"\"\"\n","    sequence1: str\n","        the first protein sequence\n","    sequence2: str\n","        the second protein sequence\n","\n","    S: blosum._blosum.BLOSUM matrix\n","        the substitution matrix to score matches and mismatches\n","    gap_penalty: int, optional (default=-8)\n","        the gap penalty to use (note: does not support affine gap penalty)\n","    \"\"\"\n","    # initialising the table\n","    n = len(sequence1)\n","    m = len(sequence2)\n","\n","    score = np.zeros((n+1, m+1), dtype=int)\n","    traceback = np.zeros((n+1, m+1), dtype=int)\n","    # ensures F(0,0) = 0\n","\n","    # Filling the first row - F(i,0) = F(i−1,0) − d\n","    for i in range(1, n+1):\n","        score[i, 0] = i * gap_penalty\n","        traceback[i, 0] = 1\n","\n","    # Filling the first column - F(0,j) = F(0,j-1) − d\n","    for j in range(1, m+1):\n","        score[0, j] = j * gap_penalty\n","        traceback[0, j] = 2\n","\n","    # getting max score\n","    for i in range(1, n+1):\n","        for j in range(1, m+1):\n","            # calculating for match.mismatch - F(i−1,j−1)+s(xi​,yj​)\n","            match = score[i-1, j-1] + S[sequence1[i-1]][sequence2[j-1]]\n","\n","            # calculating for deletion -  F(i,0) = F(i−1,j)−d\n","            delete = score[i-1, j] + gap_penalty\n","\n","            # calculating for insertion -  F(i,j) = F(i,j−1) − d\n","            insert = score[i, j-1] + gap_penalty\n","\n","            score[i, j] = max(match, delete, insert)\n","\n","            # Keeping a track of match/insertion/deletion\n","            if score[i, j] == match:\n","                traceback[i, j] = 0\n","            elif score[i, j] == delete:\n","                traceback[i, j] = 1\n","            else:\n","                traceback[i, j] = 2\n","\n","    align1 = \"\"\n","    align2 = \"\"\n","    i, j = n, m\n","\n","    # Aligning final match sequences\n","    while i > 0 or j > 0:\n","        if traceback[i, j] == 0:\n","            align1 = sequence1[i-1] + align1\n","            align2 = sequence2[j-1] + align2\n","            i -= 1\n","            j -= 1\n","        elif traceback[i, j] == 1:\n","            align1 = sequence1[i-1] + align1\n","            align2 = \"-\" + align2\n","            i -= 1\n","        else:\n","            align1 = \"-\" + align1\n","            align2 = sequence2[j-1] + align2\n","            j -= 1\n","\n","    return align1, align2, int(score[n, m]), score, traceback # Replace it and Complete the function"]},{"cell_type":"code","source":["def visualize_traceback(score, traceback, seq1_labels=None, seq2_labels=None, cell_w=7):\n","    \"\"\"\n","    Print a grid of scores with an arrow in each cell.\n","\n","    Args:\n","        score (array-like): score matrix shape (n+1, m+1)\n","        traceback (array-like): traceback matrix same shape (0=diag,1=up,2=left)\n","        seq1_labels (str or list, optional): labels for rows (original seq1). If None, blanks used.\n","        seq2_labels (str or list, optional): labels for cols (original seq2). If None, blanks used.\n","        cell_w (int): width of each printed cell\n","    \"\"\"\n","    import numpy as np\n","\n","    score = np.asarray(score)\n","    tb = np.asarray(traceback)\n","    rows, cols = score.shape  # rows = n+1, cols = m+1\n","    n = rows - 1\n","    m = cols - 1\n","\n","    # prepare labels (index 0 is the gap/corner)\n","    def to_labels(x, length):\n","        if x is None:\n","            return [' '] * length\n","        if isinstance(x, str):\n","            x = list(x)\n","        x = list(x)\n","        # truncate or pad with '-' to required length\n","        if len(x) > length:\n","            return x[:length]\n","        return x + ['-'] * (length - len(x))\n","\n","    row_labels = to_labels(seq1_labels, n)\n","    col_labels = to_labels(seq2_labels, m)\n","\n","    arrow = {0: '↖', 1: '↑', 2: '←'}\n","\n","    # header: corner then col labels (j=1..m)\n","    corner = ' ' * cell_w\n","    header = corner\n","    for j in range(1, m+1):\n","        header += f\"{col_labels[j-1]:^{cell_w}}\"\n","    print(header)\n","\n","    # rows: i=0..n\n","    for i in range(0, n+1):\n","        left_label = ' ' if i == 0 else row_labels[i-1]\n","        row_str = f\"{left_label:^{cell_w}}\"\n","        for j in range(0, m+1):\n","            sc = int(score[i, j])\n","            tb_code = int(tb[i, j]) if (0 <= i < rows and 0 <= j < cols) else None\n","            arr = arrow.get(tb_code, ' ')\n","            # show number and arrow; adjust formatting if numbers become wide\n","            cell = f\"{sc:>{cell_w-2}}{arr:>2}\"\n","            row_str += cell\n","        print(row_str)"],"metadata":{"id":"vp3elET9peDj","executionInfo":{"status":"ok","timestamp":1770763247080,"user_tz":300,"elapsed":39,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"}}},"id":"vp3elET9peDj","execution_count":173,"outputs":[]},{"cell_type":"code","execution_count":174,"id":"24377364-59ff-44d2-9bd6-0a05c76a9697","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":32,"status":"ok","timestamp":1770763247391,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"24377364-59ff-44d2-9bd6-0a05c76a9697","outputId":"d240e8b6-10d7-4b93-ce56-2730a8f53212"},"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment:\n","HEAGAWGHE-E\n","--P-AW-HEAE\n","Final score: -5\n","          P      A      W      H      E      A      E   \n","           0 ↖   -8 ←  -16 ←  -24 ←  -32 ←  -40 ←  -48 ←  -56 ←\n","   H      -8 ↑   -3 ↖  -10 ↖  -18 ←  -16 ↖  -24 ←  -32 ←  -40 ←\n","   E     -16 ↑  -10 ↖   -4 ↖  -12 ←  -18 ↖  -10 ↖  -18 ←  -26 ↖\n","   A     -24 ↑  -17 ↖   -5 ↖   -7 ↖  -14 ↖  -18 ↑   -5 ↖  -13 ←\n","   G     -32 ↑  -25 ↑  -13 ↑   -9 ↖  -10 ↖  -17 ↖  -13 ↑   -8 ↖\n","   A     -40 ↑  -33 ↖  -20 ↖  -16 ↖  -11 ↖  -11 ↖  -12 ↖  -14 ↖\n","   W     -48 ↑  -41 ↑  -28 ↑   -9 ↖  -17 ←  -15 ↖  -14 ↖  -16 ↖\n","   G     -56 ↑  -49 ↑  -36 ↑  -17 ↑  -12 ↖  -20 ↖  -15 ↖  -17 ↖\n","   H     -64 ↑  -57 ↑  -44 ↑  -25 ↑   -9 ↖  -12 ↖  -20 ←  -15 ↖\n","   E     -72 ↑  -65 ↑  -52 ↑  -33 ↑  -17 ↑   -3 ↖  -11 ←  -14 ↖\n"]}],"source":["sequence1 = \"HEAGAWGHEE\"\n","sequence2 = \"PAWHEAE\"\n","blosum_80_matrix = bl.BLOSUM(80)\n","\n","align1, align2, final_score, score_mat, tb = NW(sequence1, sequence2, S=blosum_80_matrix)\n","print(\"Alignment:\")\n","print(align1)\n","print(align2)\n","print(\"Final score:\", final_score)\n","\n","visualize_traceback(score_mat, tb, seq1_labels=sequence1, seq2_labels=sequence2)"]},{"cell_type":"markdown","id":"2cfb805a-1892-4cbd-bbbd-d7dab931d3c8","metadata":{"id":"2cfb805a-1892-4cbd-bbbd-d7dab931d3c8"},"source":["### 04B: Now run the same code using a gap penalty of -10. (20 points)\n","\n","You should get different results. Explain why, including discussing the total alignment score and the match/mismatch scores of the relevant amino acids (1-2 sentences).\n","\n"]},{"cell_type":"markdown","id":"So2CNk623vFs","metadata":{"id":"So2CNk623vFs"},"source":["*Answer*: A higher gap penalty makes gaps more costly, causing the algorithm to prefer mismatches in some positions. This changes the dynamic programming scores, traceback path, and ultimately final alignment."]},{"cell_type":"code","execution_count":163,"id":"9e25a7c8-927f-4c3f-8ba8-6a44736b6ab8","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1770762628283,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"9e25a7c8-927f-4c3f-8ba8-6a44736b6ab8","outputId":"9f8365d9-4a7a-4dd4-f8db-d029adb248a7"},"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment:\n","HEAGAWGHEE\n","--P-AWHEAE\n","Final score: -13\n","          P      A      W      H      E      A      E   \n","           0 ↖  -10 ←  -20 ←  -30 ←  -40 ←  -50 ←  -60 ←  -70 ←\n","   H     -10 ↑   -3 ↖  -12 ↖  -22 ←  -22 ↖  -32 ←  -42 ←  -52 ←\n","   E     -20 ↑  -12 ↖   -4 ↖  -14 ←  -22 ↖  -16 ↖  -26 ←  -36 ↖\n","   A     -30 ↑  -21 ↖   -7 ↖   -7 ↖  -16 ↖  -23 ↖  -11 ↖  -21 ←\n","   G     -40 ↑  -31 ↑  -17 ↑  -11 ↖  -10 ↖  -19 ↖  -21 ↑  -14 ↖\n","   A     -50 ↑  -41 ↖  -26 ↖  -20 ↖  -13 ↖  -11 ↖  -14 ↖  -22 ↖\n","   W     -60 ↑  -51 ↑  -36 ↑  -15 ↖  -23 ↖  -17 ↖  -14 ↖  -18 ↖\n","   G     -70 ↑  -61 ↑  -46 ↑  -25 ↑  -18 ↖  -26 ↖  -17 ↖  -17 ↖\n","   H     -80 ↑  -71 ↑  -56 ↑  -35 ↑  -17 ↖  -18 ↖  -27 ↑  -17 ↖\n","   E     -90 ↑  -81 ↑  -66 ↑  -45 ↑  -27 ↑  -11 ↖  -19 ↖  -21 ↖\n","   E    -100 ↑  -91 ↑  -76 ↑  -55 ↑  -37 ↑  -21 ↖  -12 ↖  -13 ↖\n"]}],"source":["sequence1 = \"HEAGAWGHEE\"\n","sequence2 = \"PAWHEAE\"\n","blosum_80_matrix = bl.BLOSUM(80)\n","\n","align1, align2, final_score, score_mat, tb = NW(sequence1, sequence2, S=blosum_80_matrix, gap_penalty=-10)\n","print(\"Alignment:\")\n","print(align1)\n","print(align2)\n","print(\"Final score:\", final_score)\n","\n","# visualize using the ORIGINAL sequences (sequence1 down rows, sequence2 across columns)\n","visualize_traceback(score_mat, tb, seq1_labels=sequence1, seq2_labels=sequence2)"]},{"cell_type":"markdown","id":"7d061a41-e60b-454a-ab69-7cd12ad175e7","metadata":{"id":"7d061a41-e60b-454a-ab69-7cd12ad175e7"},"source":["## 05. Construct an alignment of the S protein against the S protein from MERS (5 points)"]},{"cell_type":"markdown","id":"7250b8cd-78e3-4653-a8c0-c58ad645bac4","metadata":{"id":"7250b8cd-78e3-4653-a8c0-c58ad645bac4"},"source":["Sars-Cov-2, the virus that causes covid-19, is related to other viruses including the one that causes Middle East respiratory syndrome (MERS). This can be seen by aligning the sequences of the S proteins from the two viruses. Read in the MERS_S_protein.faa sequence from the data folder, and use your Needleman-Wunsch code to align the two sequences (using a gap penalty of -8) and report their alignment score."]},{"cell_type":"code","execution_count":164,"id":"e2b42927-422e-420a-8b6e-7ca558a60b3e","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":32,"status":"ok","timestamp":1770762636214,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"e2b42927-422e-420a-8b6e-7ca558a60b3e","outputId":"10da8aac-9478-41f3-f2b8-65f61d56d0c4"},"outputs":[{"output_type":"stream","name":"stdout","text":["MERS_S spike length: 1353\n","MERS_S spike: MIHSVFLLMFLLTPTESYVDVGPDSVKSACIEVDIQQTFFDKTWPRPIDVSKADGIIYPQGRTYSNITITYQGLFPYQGDHGDMYVYSAGHATGTTPQKLFVANYSQDVKQFANGFVVRIGAAANSTGTVIISPSTSATIRKIYPAFMLGSSVGNFSDGKMGRFFNHTLVLLPDGCGTLLRAFYCILEPRSGNHCPAGNSYTSFATYHTPATDCSDGNYNRNASLNSFKEYFNLRNCTFMYTYNITEDEILEWFGITQTAQGVHLFSSRYVDLYGGNMFQFATLPVYDTIKYYSIIPHSIRSIQSDRKAWAAFYVYKLQPLTFLLDFSVDGYIRRAIDCGFNDLSQLHCSYESFDVESGVYSVSSFEAKPSGSVVEQAEGVECDFSPLLSGTPPQVYNFKRLVFTNCNYNLTKLLSLFSVNDFTCSQISPAAIASNCYSSLILDYFSYPLSMKSDLSVSSAGPISQFNYKQSFSNPTCLILATVPHNLTTITKPLKYSYINKCSRFLSDDRTEVPQLVNANQYSPCVSIVPSTVWEDGDYYRKQLSPLEGGGWLVASGSTVAMTEQLQMGFGITVQYGTDTNSVCPKLEFANDTKIASQLGNCVEYSLYGVSGRGVFQNCTAVGVRQQRFVYDAYQNLVGYYSDDGNYYCLRACVSVPVSVIYDKETKTHATLFGSVACEHISSTMSQYSRSTRSMLKRRDSTYGPLQTPVGCVLGLVNSSLFVEDCKLPLGQSLCALPDTPSTLTPRSVRSVPGEMRLASIAFNHPIQVDQLNSSYFKLSIPTNFSFGVTQEYIQTTIQKVTVDCKQYVCNGFQKCEQLLREYGQFCSKINQALHGANLRQDDSVRNLFASVKSSQSSPIIPGFGGDFNLTLLEPVSISTGSRSARSAIEDLLFDKVTIADPGYMQGYDDCMQQGPASARDLICAQYVAGYKVLPPLMDVNMEAAYTSSLLGSIAGVGWTAGLSSFAAIPFAQSIFYRLNGVGITQQVLSENQKLIANKFNQALGAMQTGFTTTNEAFHKVQDAVNNNAQALSKLASELSNTFGAISASIGDIIQRLDVLEQDAQIDRLINGRLTTLNAFVAQQLVRSESAALSAQLAKDKVNECVKAQSKRSGFCGQGTHIVSFVVNAPNGLYFMHVGYYPSNHIEVVSAYGLCDAANPTNCIAPVNGYFIKTNNTRIVDEWSYTGSSFYAPEPITSLNTKYVAPQVTYQNISTNLPPPLLGNSTGIDFQDELDEFFKNVSTSIPNFGSLTQINTTLLDLTYEMLSLQQVVKALNESYIDLKELGNYTYYNKWPWYIWLGFIAGLVALALCVFFILCCTGCGTNCMGKLKCNRCCDRYEEYDLEPHKVHVH\n"]}],"source":["for record in SeqIO.parse(\"/content/drive/MyDrive/690U_assignments/SARS_data/MERS_S_protein.faa\", \"fasta\"):\n","      mers_s_spike = str(record.seq)\n","\n","print(\"MERS_S spike length:\", len(mers_s_spike))\n","print(\"MERS_S spike:\", mers_s_spike)"]},{"cell_type":"code","execution_count":165,"id":"7GRS4yhc0Jgy","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":19,"status":"ok","timestamp":1770762636235,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"7GRS4yhc0Jgy","outputId":"45fceebe-85da-40f6-fb1d-a6cc6547b4f7"},"outputs":[{"output_type":"stream","name":"stdout","text":["NC_045512.2 29903\n","Gene:  ID: NC_045512.2\n","Name: NC_045512.2\n","Description: NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n","Number of features: 0\n","Seq('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGT...AAA')\n"]}],"source":["for gene_record in SeqIO.parse(\"/content/ncbi_dataset/data/genomic.fna\", \"fasta\"):\n","    print(\"%s %i\" % (gene_record.id, len(gene_record)))\n","    print(\"Gene: \", gene_record)"]},{"cell_type":"code","execution_count":166,"id":"oiPuz7co0M0l","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":12,"status":"ok","timestamp":1770762636257,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"oiPuz7co0M0l","outputId":"78b2658b-f720-42ac-e019-19d31b08bd3c"},"outputs":[{"output_type":"stream","name":"stdout","text":["MIHSVFLLMFLLTPTESYVDVGPDSVKSACIEVDIQQTFFDKTWPRPIDVSKADGIIYPQGRTYSNITITYQGLFPYQGDHGDMYVYSAGHATGTTPQKLFVANYSQDVKQFANGFVVRIGAAANSTGTVIISPSTSATIRKIYPAFMLGSSVGNFSDGKMGRFFNHTLVLLPDGCGTLLRAFYCILEPRSGNHCPAGNSYTSFATYHTPATDCSDGNYNRNASLNSFKEYFNLRNCTFMYTYNITEDEILEWFGITQTAQGVHLFSSRYVDLYGGNMFQFATLPVYDTIKYYSIIPHSIRSIQSDRKAWAAFYVYKLQPLTFLLDFSVDGYIRRAIDCGFNDLSQLHCSYESFDVESGVYSVSSFEAKPSGSVVEQAEGVECDFSPLLSGTPPQVYNFKRLVFTNCNYNLTKLLSLFSVNDFTCSQISPAAIASNCYSSLILDYFSYPLSMKSDLSVSSAGPISQFNYKQSFSNPTCLILATVPHNLTTITKPLKYSYINKCSRFLSDDRTEVPQLVNANQYSPCVSIVPSTVWEDGDYYRKQLSPLEGGGWLVASGSTVAMTEQLQMGFGITVQYGTDTNSVCPKLEFANDTKIASQLGNCVEYSLYGVSGRGVFQNCTAVGVRQQRFVYDAYQNLVGYYSDDGNYYCLRACVSVPVSVIYDKETKTHATLFGSVACEHISSTMSQYSRSTRSMLKRRDSTYGPLQTPVGCVLGLVNSSLFVEDCKLPLGQSLCALPDTPSTLTPRSVRSVPGEMRLASIAFNHPIQVDQLNSSYFKLSIPTNFSFGVTQEYIQTTIQKVTVDCKQYVCNGFQKCEQLLREYGQFCSKINQALHGANLRQDDSVRNLFASVKSSQSSPIIPGFGGDFNLTLLEPVSISTGSRSARSAIEDLLFDKVTIADPGYMQGYDDCMQQGPASARDLICAQYVAGYKVLPPLMDVNMEAAYTSSLLGSIAGVGWTAGLSSFAAIPFAQSIFYRLNGVGITQQVLSENQKLIANKFNQALGAMQTGFTTTNEAFHKVQDAVNNNAQALSKLASELSNTFGAISASIGDIIQRLDVLEQDAQIDRLINGRLTTLNAFVAQQLVRSESAALSAQLAKDKVNECVKAQSKRSGFCGQGTHIVSFVVNAPNGLYFMHVGYYPSNHIEVVSAYGLCDAANPTNCIAPVNGYFIKTNNTRIVDEWSYTGSSFYAPEPITSLNTKYVAPQVTYQNISTNLPPPLLGNSTGIDFQDELDEFFKNVSTSIPNFGSLTQINTTLLDLTYEMLSLQQVVKALNESYIDLKELGNYTYYNKWPWYIWLGFIAGLVALALCVFFILCCTGCGTNCMGKLKCNRCCDRYEEYDLEPHKVHVH\n","MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAIHVSGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLGVYYHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINLVRDLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPCNGVEGFNCYFPLQSYGFQPTNGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIADTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQDVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSYECDIPIGAGICASYQTQTNSPRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT_\n","lengths:  1274 1274\n"]}],"source":["reference_spike, reference_spike_len = translate_seq(gene_record, translation_table)\n","print(mers_s_spike)\n","print(reference_spike)\n","print(\"lengths: \", mers_s_spike_len, reference_spike_len)"]},{"cell_type":"code","execution_count":169,"id":"cBar_Wiph3DK","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"elapsed":9748,"status":"ok","timestamp":1770762749806,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"},"user_tz":300},"id":"cBar_Wiph3DK","outputId":"1daa1cb9-c786-4985-c3a2-7e4f2198750f","collapsed":true},"outputs":[{"output_type":"stream","name":"stdout","text":["Alignment:\n","MF--VFLVLLPLVSSQC-VNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQD-LFLP---FFSNVTW-FHAIH-VSGTNGTK-RFD-NPVLPFN-DGVYFA--STEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVI-KVC-EFQF---C-N--DPFLGVYY-H-----KNN-KSWMESEFRVYS--SANNC-T-FEYVS-QPFLM-DLEGKQGNF-KN--LREF-VFKNI-D-GYFKIYS-KHTPI-N-L-VRDLPQGFSALEP-LVDLPIGINITRFQTL-L-ALHRSY-LTPGD-SSSGWTAGA-AAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVI-AWNSNNLDSKVGG-NYNYLYRLFR-KSNLK-PFERDISTEIYQ-AGS-TPCNGVE-G--FNCYF-PLQSYGFQPTNGVGYQPYRVVVLSFEL-L-HAPAT--VCGPK-K-S--TNLV-K-NKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRD-IADTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQDVNCTEVPVAI-HADQLTPT-WRVYSTGSNVFQTRAGCLIGAEHVNNSYE-CDIPIGAGICASYQT-QTNSPRRARSV-ASQSI--IAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGG-FNFSQILP-DPSKPSK--RSFIEDLLFNKVTLADAGFIKQYGDCL--GDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITS-GWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGK-AH-F-PREGVFV-SNGT----HWFVTQRNFYEPQIITTDNT-FVSGNCDVVIGIVNNTVYDP-LQPELD-SFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVM-VTIMLCCMTSC-CSCLKGCCSCGSCC-KFDEDDSEPVLKGVKLHYT\n","MIHSVFLLMFLLTPTESYVDV-GPDSVKSACIEVDIQQTFF-DKTWPRPIDVSKADGIIYPQGRTYSNITITYQGLFPYQGDHGDMYVYSAGHATGTTPQKLFVANYSQDVKQFANGFVVRIGAAANSTGTVIISPSTSATIRKIYPAFMLGSSVGNFSDGKMGRFFNHTLVLLPDGCGTLLRAFYCILEPRSGNHCPAGNSYTSFATYHTPATDCSDGNYNRNASLNSFKEYFNLRNCTFMYTYNITEDEILEWFGITQTAQGVHLFSSRYVDL-YGGNMFQFATLPVYDTIKYYSIIPHSIRSIQSDRKAWAAFYVYKLQPLTFLLDFSVDGYIRRAIDCGFNDLSQLHCSYESFDVESGVYSVSSFEAKPSGSVVEQAEGVE-CDFSPLLSGTP-PQVYNFKRLVFTNCNYNLTKLLSLFSVNDFTCSQISPAAIASNCYSSLILDYFSYPLSMKSDLSVSSAGPISQFNYKQSFSNPTCLILATVPHNLTTITKPLKYSYINKCSRFLSDDRTEVPQLVNANQYSPCVSIVPSTVWEDGDYYRKQLSPLEGGGWLVASGSTVAMTEQLQMGFGITVQYGTDTNSVC-PKLEFANDTKIASQLGNCVEYSLYGVSGRGVFQNCTAVGVRQQRFVYDAYQNLVGYYSDDGNYYCL--RACVSVPVSVIY-DKETKTH-ATLFGSVACEHISSTMSQYSRSTRSMLKRRDSTYGPLQTPVGCVLGLVNSSLFVEDCKLPLGQSLCALPDTPSTLTPRSVRSVPGEMRLASIAFNHPIQV-DQLNSSYFKLSIPTNFSFGVTQEYIQTTIQKVTVDCKQYVCNGFQKCEQLLREYGQFCSKINQALHGANLRQDDSVRNLFASVKSSQSSPIIPGFGGDFNLTLLEPVSISTGSRSARSAIEDLLFDKVTIADPGYMQGYDDCMQQGPASARDLICAQYVAGYKVLPPLMDVNMEAAYTSSLL-GSIAGVGWTAGLSSFAAIPFAQSIFYRLNGVGITQQVLSENQKLIANKFNQALGAMQTGFTTTNEAFHKVQDAVNNNAQALSKLASELSNTFGAISASIGDIIQRLDVLEQDAQIDRLINGRLTTLNAFVAQQLVRSESAALSAQLAKDKVNECVKAQSKRSGFCGQGTHIVSFVVNAPNGLYFMHVGYYPSNHIEVVSAYGLCDAANPTNCIAPVNGYFIKTNNTRIVDEWSYTGSSFYAPEPITSLNTKYVAPQV-TYQNISTN-LPPPLLGNSTGIDFQDELDEFFKNVSTSIPNFGSLTQINTTLLDLTYEMLSLQQVVKALNESYIDLKELGNYTYYNKWPWYIWLGFIAGLVALALCVFFILCC-TGCGTNCM-GKLKCNRCCDRYEEYDLEP-HK-VHVH--\n","Final score: 989\n"]}],"source":["# sequence matching\n","\n","blosum_80_matrix = bl.BLOSUM(80)\n","align1, align2, final_score, score_mat, tb = NW(protein_seq[:-1], mers_s_spike, S=blosum_80_matrix)\n","print(\"Alignment:\")\n","print(align1)\n","print(align2)\n","print(\"Final score:\", final_score)\n","\n","# visualize using the ORIGINAL sequences (sequence1 down rows, sequence2 across columns)\n","# visualize_traceback(score_mat, tb, seq1_labels=sequence1, seq2_labels=sequence2)"]},{"cell_type":"markdown","id":"7125366d-bc3e-488f-97e4-a9ad02b034e0","metadata":{"id":"7125366d-bc3e-488f-97e4-a9ad02b034e0"},"source":["While it may be challenging to see the similarity in these sequences by eye, I encourage you to view figure 1 in this paper: https://www.biorxiv.org/content/10.1101/2020.04.17.047548v1.full which shows the structural similarity in the S protein between the two sequences. Alignments like this were used in the early days of the pandemic to establish that covid-19 was related to previously observed respiratory viruses such as MERS."]},{"cell_type":"code","execution_count":null,"id":"Btggkaujy9oF","metadata":{"id":"Btggkaujy9oF","executionInfo":{"status":"aborted","timestamp":1770762644926,"user_tz":300,"elapsed":8774,"user":{"displayName":"Helly Dhamesha","userId":"06774192143955883616"}}},"outputs":[],"source":[]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.1"}},"nbformat":4,"nbformat_minor":5}